_posts/2020-05-25-clojurescript-web-serverwith-macchiato-shadowcljs-and-reitit.md

```diff
@@ -0,0 +1,329 @@
+     :post {:parameters {:body {:my-body string?}}
+            :handler    (fn [request respond _]
+                          (respond {:status 200 :body {:message (str "Hello: " (-> request :parameters :body :my-body))}}))}}]
+   ["/bad-response-bug"
+       {:get  {:parameters {:query {:name string?}}
+               :responses  {200 {:body {:message string?}}}
+               :handler    (fn [request respond _]
+                             (respond {:status 200 :body {:messag (str "Hello: " (-> request :parameters :query :name))}}))}}]])
+
+(defn wrap-coercion-exception
+  "Catches potential synchronous coercion exception in middleware chain"
+  [handler]
+  (fn [request respond _]
+    (try
+      (handler request respond _)
+      (catch :default e
+        (let [exception-type (:type (.-data e))]
+          (cond
+            (= exception-type :reitit.coercion/request-coercion)
+            (respond {:status 400
+                      :body   {:message "Bad Request"}})
+
+            (= exception-type :reitit.coercion/response-coercion)
+            (respond {:status 500
+                      :body   {:message "Bad Response"}})
+            :else
+            (respond {:status 500
+                      :body   {:message "Truly internal server error"}})))))))
+
+(defn wrap-body-to-params
+  [handler]
+  (fn [request respond raise]
+    (handler (-> request
+                 (assoc-in [:params :body-params] (:body request))
+                 (assoc :body-params (:body request))) respond raise)))
+
+(def app
+  (ring/ring-handler
+    (ring/router
+      [routes]
+      {:data {:middleware [params/wrap-params
+                           #(rf/wrap-restful-format % {:keywordize? true})
+                           wrap-body-to-params
+                           wrap-coercion-exception
+                           rrc/coerce-request-middleware
+                           rrc/coerce-response-middleware]}})
+    (ring/create-default-handler)))
+
+
+(defn server []
+  (info "Hey I am running now!")
+  (let [host "127.0.0.1"
+        port 3000]
+    (http/start
+      {:handler    app
+       :host       host
+       :port       port
+       :on-success #(info "macchiato-test started on" host ":" port)})))
+```
+## Middlewares 
+
+First two middlewares are needed for parameter wrapping. `params/wrap-params` does query params parsing and reads possible
+body of a HTTP request. `rf/wrap-restful-format` does encoding/decoding depending on Content-Type of the request. `wrap-body-to-params`
+is a new middleware that I made because Reitit expects body params to be in `body-params` named map in the Ring request.
+
+`wrap-coercion-exception` is a middleware which catches request and response coercion errors and returns 400 or 500 level 
+error messages. In real development at least 400 error should include also some information why the requests are rejected. 
```

Comment:
What does "real development" mean? "In a real application", maybe?

"should also include information on why…"